pkg/middleware/stored_session.go

```diff
@@ -252,3 +262,58 @@ func (s *storedSessionLoader) validateSession(ctx context.Context, session *sess
 
 	return nil
 }
+
+// forceRefreshSession attempts to load a session as identified by the request cookies.
+// If no session is found, the request will be passed to the next handler.
+// If a session was loaded by a previous handler, it will not be refreshed.
+func (s *storedSessionLoader) forceRefreshSession(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		scope := middlewareapi.GetRequestScope(req)
+
+		// If refreshPeriod isn't defined then forward request to the next handler
+		if s.refreshPeriod <= time.Duration(0) {
+			next.ServeHTTP(rw, req)
+			return
+		}
+
+		session, err := s.store.Load(req)
+		if err != nil {
+			s.handleRefreshError(rw, req, next, err)
+			return
+		}
+
+		if session == nil {
+			next.ServeHTTP(rw, req)
+			return
+		}
+
+		err = s.refreshSessionIfNeeded(rw, req, session, true)
+		if err != nil {
+			s.handleRefreshError(rw, req, next, err)
+			return
+		}
+
+		// Add the session to the scope if it was found
+		scope.Session = session
+		next.ServeHTTP(rw, req)
+	})
+}
+
+func (s *storedSessionLoader) handleRefreshError(rw http.ResponseWriter, req *http.Request, next http.Handler, err error) {
```

Comment:
Could we/should we use this same function to handle the failures in the other place that the `refreshSessionIfneeded` function is called?

Reply:
Where else would you want it used?  I think it's used everywhere I touched the codebase.

Reply:
In the existing code, we call `getValidatedSession` which does something similar to what you are doing when you call `refreshSessionIfNeeded`, and then it has similar error handling.

I would expect we can use this function in both locations to have the same refresh error handling in both places without duplication of the code